doc/designs/quic-design/dgram-api.md

```diff
@@ -0,0 +1,322 @@
+      (see discussion below), We would need to define our own structures in this
+      case (a “polyfill” approach).
+
+  - If we use our own structures:
+
+    - We would need to translate these structures during every call.
+
+      But we would need to have storage inside the BIO_dgram for *m* `struct
+      msghdr`, *m\*v* iovecs, etc. Since we want to support multithreaded use
+      these allocations probably will need to be on the stack, and therefore
+      must be limited.
+
+      Limiting *m* isn't a problem, because `sendmmsg` returns the number
+      of messages sent, so the existing semantics we are trying to match
+      lets us just send or receive fewer messages than we were asked to.
+
+      However, it does seem like we will need to limit *v*, the number of iovecs
+      per message. So what limit should we give to *v*, the number of iovecs? We
+      will need a fixed stack allocation of OS iovec structures and we can
+      allocate from this stack allocation as we iterate through the `BIO_msghdr`
+      we have been given. So in practice we could just only send messages
+      until we reach our iovec limit, and then return.
+
+      For example, suppose we allocate 64 iovecs internally:
+
+      ```c
+      struct iovec vecs[64];
+      ```
+
+      If the first message passed to a call to `BIO_writem` has 64 iovecs
+      attached to it, no further messages can be sent and `BIO_writem`
+      returns 1.
+
+      If three messages are sent, with 32, 32, and 1 iovecs respectively,
+      the first two messages are sent and `BIO_writem` returns 2.
+
+      So the only important thing we would need to document in this API
+      is the limit of iovecs on a single message; in other words, the
+      number of iovecs which must not be exceeded if a forward progress
+      guarantee is to be made. e.g. if we allocate 64 iovecs internally,
+      `BIO_writem` with a single message with 65 iovecs will never work
+      and this becomes part of the API contract.
+
+      Obviously these quantities of iovecs are unrealistically large.
+      iovecs are small, so we can afford to set the limit high enough
+      that it shouldn't cause any problems in practice. We can increase
+      the limit later without a breaking API change, but we cannot decrease
+      it later. So we might want to start with something small, like 8.
+
+- We also need to decide what to do for OSes which don't support at least
+  `sendmsg`/`recvmsg`.
+
+  - Don't provide these functions and require all users of these functions to
+    have an alternate code path which doesn't rely on them?
+
+    - Not providing these functions on OSes that don't support
+      at least sendmsg/recvmsg is a simple solution but adds
+      complexity to code using BIO_dgram. (Though it does communicate
+      to code more realistic performance expectations since it
+      knows when these functions are actually available.)
```

Comment:
It feels unappealing to have a "not provide a function on some OSes" API, though sometimes it is unavoidable.  (E.g., we have a separate configure option for `ec_nistp_64_gcc_128` since the `int128_t` type is not available everywhere.  I thought there was also something socket- or stdio-related that was unavailable on Windows but am not finding it now.)  I would tend to favor the ability to write portable code over some "purity" argument that says an mmsg function needs to be faster than repeatedly calling the regular send/recvmsg functions.